drake/solvers/Optimization.h

```diff
@@ -108,46 +116,123 @@ namespace Drake {
    */
   class Constraint {
   public:
-    Constraint(const VariableList& vars) : variable_list(vars) {}
+    Constraint(const VariableList& vars, size_t num_constraints) : variable_list(vars), lower_bound(num_constraints), upper_bound(num_constraints) {
+      assert(lower_bound.size() == upper_bound.size() && "Lower bound and upper bound must be the same size");
+      lower_bound.setConstant(-std::numeric_limits<double>::infinity());
+      upper_bound.setConstant(std::numeric_limits<double>::infinity());
+    }
+
+    template <typename DerivedLB, typename DerivedUB>
+    Constraint(const VariableList& vars, const Eigen::MatrixBase<DerivedLB>& lb, const Eigen::MatrixBase<DerivedUB>& ub) : variable_list(vars), lower_bound(lb), upper_bound(ub)
+    {}
     virtual ~Constraint() {}
 
+    virtual void eval(const Eigen::Ref<const Eigen::VectorXd>& x, Eigen::VectorXd& y) const = 0;
+    virtual void eval(const Eigen::Ref<const TaylorVecXd>& x, TaylorVecXd& y) const = 0;  // move this to DifferentiableConstraint derived class if/when we need to support non-differentiable functions
+
+    virtual Eigen::VectorXd getLowerBound() const { return lower_bound; }
+    virtual Eigen::VectorXd getUpperBound() const { return upper_bound; }
```

Comment:
actually, i revisited this and could make things const & at least for now.  but it does add a strong  constraint on the derived classes.